### ISSUE TITLE
lattice sync probs

### ISSUE THREAD

**Issue by kasperbauer** [kasperbauer/metrix#20]
Lattice gets out of sync, when setting the divisions with `pattern:set_division()` on an existing pattern.

**Comment by kasperbauer** [comment]
```
lattice = include('lib/lattice')

divisions = {
    [1] = 1 / 1,
    [2] = 1 / 2,
    [3] = 1 / 4,
    [4] = 1 / 8
}

function init()
    lat = lattice:new()
    p1 = lat:new_pattern({
        action = function(t)
            print(1, t)
        end,
        division = 1 / 8
    })
    p2 = lat:new_pattern({
        action = function(t)
            print(2, t)
        end,
        division = 1 / 8
    })
end

function key(n, z)
    if z == 1 then
        if n == 2 then
            p1:set_division(divisions[math.random(1, 4)])
            p2:set_division(divisions[math.random(1, 4)])
            print('SET DIVISION 1', p1.division)
            print('SET DIVISION 2', p2.division)
            lat:start()
            -- lat:hard_restart()
        elseif n == 3 then
            lat:stop()
        end
    end
end
```

**Comment by tyleretters** [comment]
i think all you need to do is move the hard_restart up...

```
function lattice_start()
    lat:hard_restart()  -- ... to here!
    p1:set_division(divisions[3])
    p2:set_division(divisions[4])
    print('SET DIVISION 1', p1.division)
    print('SET DIVISION 2', p2.division)
    -- used to be not here
end

```

output before moving hard_restart:
```
SET DIVISION 1	0.25
SET DIVISION 2	0.125
2	0
2	48
1	48
2	96
2	144
1	144
2	192
2	240
1	240
2	288
2	336
1	336
2	384
2	432
1	432
```


after:
```
SET DIVISION 1	0.25
SET DIVISION 2	0.125
2	48
2	96
1	96
2	144
2	192
1	192
2	240
2	288
1	288
2	336
2	384
1	384
2	432
```

**Comment by schollz** [comment]
hey @kasperbauer , I got pinged from @tyleretters about this issue and I can propose a fix. 

I think the fix needs to happen in lattice so you can either directly edit the lib version `~/norns/lua/lib/lattice.lua` or copy that into your directory and use the `include('lib/lattice') and edit it there.

the fix (I think) should be to change this

```lua
function Pattern:set_division(n)
   self.division = n
end
```
to 

```lua
function Pattern:set_division(n)
   self.division = n
   self.phase_end = self.division * self.ppqn * self.meter
end
```

if that works for you, please let us know and a pr can be made to the lattice in the norns core.

**Comment by kasperbauer** [comment]
Hey there @schollz @tyleretters 
Thanks for your help, i think now I'm getting the hang of how lattice works :)

I had to move the phase reset up to lattice, as the pattern class doesn't know the ppqn and meter values.
c7c9e16456b993200357129e7feba58d8fa9d852

I also added a method for resetting lattice without restarting, as i also wanted to change the division beforehand, without running lattice.
7c71fd9b2ad8628776a7a2b6a535d1fa8b6226f9

Take a look, if you're fine with it I'd be happy to provide a PR in norns core.
